File: Java/algorithms/sorting/QuickSort.java
/**
 * Algorithm: Quick Sort
 * Description: Efficient divide-and-conquer sorting algorithm using partitioning
 * Time Complexity: O(n log n) average case, O(n²) worst case
 * Space Complexity: O(log n) for recursion stack
 *
 * @author Pasan Sulakshana Rathnayake
 * @version 1.0
 * @since 2025-10-03
 */

import java.util.*;

public class QuickSort {

    /**
     * Sorts an array using quick sort algorithm.
     *
     * @param arr the array to be sorted
     * @throws IllegalArgumentException if array is null
     */
    public static void quickSort(int[] arr) {
        if (arr == null) {
            throw new IllegalArgumentException("Array cannot be null");
        }

        if (arr.length <= 1) {
            return; // Already sorted
        }

        quickSort(arr, 0, arr.length - 1);
    }

    /**
     * Recursive quick sort implementation.
     *
     * @param arr the array to sort
     * @param low starting index
     * @param high ending index
     */
    private static void quickSort(int[] arr, int low, int high) {
        if (low < high) {
            // Partition the array and get pivot index
            int pivotIndex = partition(arr, low, high);
            
            // Recursively sort elements before and after partition
            quickSort(arr, low, pivotIndex - 1);
            quickSort(arr, pivotIndex + 1, high);
        }
    }

    /**
     * Partitions the array around a pivot element.
     *
     * @param arr the array to partition
     * @param low starting index
     * @param high ending index
     * @return the final position of the pivot
     */
    private static int partition(int[] arr, int low, int high) {
        // Choose rightmost element as pivot
        int pivot = arr[high];
        int i = low - 1; // Index of smaller element
        
        for (int j = low; j < high; j++) {
            // If current element is smaller than or equal to pivot
            if (arr[j] <= pivot) {
                i++;
                swap(arr, i, j);
            }
        }
        
        // Place pivot in correct position
        swap(arr, i + 1, high);
        return i + 1;
    }

    /**
     * Swaps two elements in the array.
     *
     * @param arr the array
     * @param i first index
     * @param j second index
     */
    private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

    /**
     * Test the quick sort implementation.
     */
    public static void main(String[] args) {
        // Test case 1: Regular array
        int[] test1 = {64, 34, 25, 12, 22, 11, 90};
        System.out.println("Original array: " + Arrays.toString(test1));
        quickSort(test1);
        System.out.println("Sorted array: " + Arrays.toString(test1));
        
        // Test case 2: Already sorted
        int[] test2 = {1, 2, 3, 4, 5};
        System.out.println("\nAlready sorted: " + Arrays.toString(test2));
        quickSort(test2);
        System.out.println("After sorting: " + Arrays.toString(test2));
        
        // Test case 3: Reverse sorted
        int[] test3 = {5, 4, 3, 2, 1};
        System.out.println("\nReverse sorted: " + Arrays.toString(test3));
        quickSort(test3);
        System.out.println("After sorting: " + Arrays.toString(test3));
        
        // Test case 4: Single element
        int[] test4 = {42};
        quickSort(test4);
        assert test4[0] == 42;
        
        // Test case 5: Empty array
        int[] test5 = {};
        quickSort(test5);
        assert test5.length == 0;
        
        System.out.println("\nAll test cases passed! ✅");
    }
}

File: Python/algorithms/sorting/quick_sort.py
"""
Quick Sort Algorithm
Time Complexity: O(n log n) average, O(n²) worst case
Space Complexity: O(log n) due to recursion stack
Author: Karanjot Singh
Date: October 2025
Hacktoberfest 2025
"""

def partition(arr, low, high):
    """
    Partition function for quick sort
    Places pivot element at correct position and arranges smaller elements to left
    and greater elements to right of pivot

    Args:
        arr: List to partition
        low: Starting index
        high: Ending index

    Returns:
        Partition index
    """
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1


def quick_sort(arr, low=0, high=None):
    """
    Quick sort implementation using divide and conquer

    Args:
        arr: List to sort
        low: Starting index (default 0)
        high: Ending index (default None, uses len(arr)-1)
    """
    if high is None:
        high = len(arr) - 1

    if low < high:
        pi = partition(arr, low, high)
        quick_sort(arr, low, pi - 1)
        quick_sort(arr, pi + 1, high)


def quick_sort_wrapper(arr):
    """
    Wrapper function that returns sorted array

    Args:
        arr: List to sort

    Returns:
        Sorted list
    """
    arr_copy = arr.copy()
    quick_sort(arr_copy)
    return arr_copy


# Test cases
if __name__ == "__main__":
    print("=== Quick Sort Algorithm ===\n")

    # Test case 1: Random array
    test1 = [64, 34, 25, 12, 22, 11, 90]
    print(f"Original array: {test1}")
    sorted1 = quick_sort_wrapper(test1)
    print(f"Sorted array:   {sorted1}")

    # Test case 2: Already sorted
    test2 = [1, 2, 3, 4, 5]
    print(f"\nOriginal array: {test2}")
    sorted2 = quick_sort_wrapper(test2)
    print(f"Sorted array:   {sorted2}")

    # Test case 3: Reverse sorted
    test3 = [9, 7, 5, 3, 1]
    print(f"\nOriginal array: {test3}")
    sorted3 = quick_sort_wrapper(test3)
    print(f"Sorted array:   {sorted3}")

    # Test case 4: Array with duplicates
    test4 = [5, 2, 8, 2, 9, 1, 5]
    print(f"\nOriginal array: {test4}")
    sorted4 = quick_sort_wrapper(test4)
    print(f"Sorted array:   {sorted4}")

    # Test case 5: Single element
    test5 = [42]
    print(f"\nOriginal array: {test5}")
    sorted5 = quick_sort_wrapper(test5)
    print(f"Sorted array:   {sorted5}")

    print("\n✅ All test cases completed!")
